### ISSUE TITLE
Missing type error if any[] is spread into a typed array

### ISSUE THREAD

**Issue by Pyrolistical** [microsoft/TypeScript#45705]
# Bug Report

When an empty array of type `any[]` is spread into a typed array containing invalid types, the type error goes away.

### 🔎 Search Terms

- any
- spread array
- missing error

### 🕗 Version & Regression Information
Tested on nightly 4.5.0-dev.20210902

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210902#code/C4TwDgpgBAChBOBnA9gOygXgFBV1APlAOQDKwArgCYSrBE56FEAqEAhgMYAWCRA3Fiwc0iYFADuyeAGtEALlgIUqANoBdTFBUNcRAMLwIEMEVwB6M1GBcAloigJ4U+23sQAHpA7AIlLGsFhVFEJNmAFNlQQdU11ASERMQAjcgBzVN8FOCQ0GIwtHSgAOhLxMIAaQv1DY1MoCygAWztEG1RUh3gneH8sIA)

### 💻 Code

```ts
type Person =
    | 'Student'
    | 'Teacher';

const works: Person[] = [
    'Creep'  // this errors as expected
]

const wat: any[] = [];

const bugged: Person[] = [
    ...wat,
    'Creep'  // missing error
]
```

### 🙁 Actual behaviour

Only one error for the `const works`: `Type '"Creep"' is not assignable to type 'Person'.`

### 🙂 Expected behaviour

Both `const works` and `const bugged` should error with `Type '"Creep"' is not assignable to type 'Person'.`
Also, I would have expected a type error on the spread of `wat`, `Type 'any[]' is not assignable to type 'Person[]'`

**Comment by MartinJohns** [comment]
By spreading `any` you end up with `any`, and `any` is assignable to anything. If you use `any` you throw type-safety out the window.

**Comment by dmitrysteblyuk** [comment]
Really sorry for off top, but can someone point me to the explanation of why `const wat = [];` will have `any[]` type even in **strict** mode. I'd expect it to be `unknown[]` or `never[]`, so that there's no implicit any. Thanks!

**Comment by MartinJohns** [comment]
@dmitrysteblyuk Because this behaviour predates the introduction of the `unknown` type. Changing this behavior would have been a massive breaking change.

**Comment by dmitrysteblyuk** [comment]
And this is too minor a thing to add a special compiler option just for this I guess? I'd love to look into this and contribute with a PR if it's worth it.

But thank you! I've been looking everywhere for a workaround or explanation.

**Comment by MartinJohns** [comment]
> And this is too minor a thing to add a special compiler option just for this I guess?

If the team always added a new compiler flag when someone asked for it I'd quickly run out of disk space for the `tsconfig.json` alone.

Besides, the type of the array is adjusting when you push items:

    const wat = [];
    wat.push(1);
    wat; // Now typed number[]

**Comment by Pyrolistical** [comment]
I don't see how this is working as intended as `bugged` is explicitly typed `const bugged: Person[]`. Shouldn't the compiler reveal this contradiction? `Cannot assign any[] to Person[]`? I would agree this is working as intended if `bugged`'s type was inferred.

**Comment by MartinJohns** [comment]
`any` is assignable to anything, and the value you assign is typed `any[]`.

**Comment by Pyrolistical** [comment]
I see, `any` is a poison pill. I encountered this situtation with ramda's R.pipe which implicitly returns any and caused this problem.

My workaround to that problem is to turn on `"noImplicitAny": true`
